# TFM project notebook

This notebook is simple a file where I write everything I do during my master thesis, on a weekly basis, since I obtained the data to analyze (11/12/2019).

### October-November

Reading articles, defining the project, writing the project proposal (you can find it on Reports folder), making other collaborations analysis within the group, learning the quality control steps and EWAS analysis of the methylion data, and beginning to understand/propose the quality control for the gene expression array data.

### Week 9-13 December 2019

#### 11/12/2019: dbGAP approval and QC gene exp papers

We have already received approval by `dbGAP` to download the data. The next step is to know how exactly download it and where to find the gene expression array files, plus the metadata (phenotype). There are many many files, so that is not an easy task to do.
I have requested access to some ports in order to download it. While I wait the approval by the head of IT department I have beginning to search for some papers where they have already used FHS gene expression data, to look for the quality control steps.

Paper 1:

> *The raw data were quantile-normalized and log2 transformed, followed by summarization using Robust Multi-array Average [34]. The gene annotations were obtained from Affymetrix NetAffx Analysis Center (version 31). We excluded transcript clusters that were not mapped to RefSeq transcripts, resulting in 17,873 distinct transcripts (17,324 distinct genes) for downstream analysis.

#### 12/12/2019

#### 13/12/2019



### Week 16-20 December 2019

